File: packages/node/src/mcp/mcp-stdio-app.ts
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { ensureLocalDaemonRunning } from '../client/daemon-lifecycle.js';
import { createThinMcpServer } from './thin-mcp-server.js';
import { logT, portLang } from '../i18n/lang.js';
import { logger, setLogLevel } from '../utils/logger.js';

let shutdownPromise: Promise<void> | null = null;

export async function startThinMcpStdio(): Promise<void> {
  if (process.argv.includes('--debug')) {
    setLogLevel('debug');
  }

  try {
    await ensureLocalDaemonRunning();
  } catch (err) {
    logger.error(err instanceof Error ? err.message : String(err));
    process.exit(1);
    return;
  }

  const mcpServer = createThinMcpServer();
  const transport = new StdioServerTransport();
  await mcpServer.connect(transport);

  logger.info(logT(portLang(), 'mcpConnectedStdio'));

  const shutdown = async () => {
    if (shutdownPromise) return;
    shutdownPromise = (async () => {
      try {
        await mcpServer.close();
      } catch (err) {
        logger.warn(logT(portLang(), 'errorClosingMcp'), err);
      }
      process.exit(0);
    })();
  };

  process.stdin.on('end', () => {
    logger.info(logT(portLang(), 'stdinClosedShuttingDown'));
    void shutdown();
  });

  process.on('SIGINT', () => void shutdown());
  process.on('SIGTERM', () => void shutdown());
}

File: packages/python/src/mcp/mcp_stdio_app.py
from __future__ import annotations

import sys

from mcp.server.stdio import stdio_server

from lionscraper.client.daemon_lifecycle import ensure_local_daemon_running
from lionscraper.i18n.lang import log_t, port_lang
from lionscraper.mcp.thin_mcp_server import create_thin_mcp_server
from lionscraper.utils.logger import logger, set_log_level


async def start_thin_mcp_stdio() -> None:
    if "--debug" in sys.argv:
        set_log_level("debug")

    try:
        await ensure_local_daemon_running()
    except Exception as err:
        logger.error(str(err))
        raise SystemExit(1) from err

    mcp_server = create_thin_mcp_server()
    init = mcp_server.create_initialization_options()
    L = port_lang()

    async with stdio_server() as (read_stream, write_stream):
        logger.info(log_t(L, "mcpConnectedStdio"))
        await mcp_server.run(read_stream, write_stream, init)

    logger.info(log_t(L, "stdinClosedShuttingDown"))
